presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java

```diff
@@ -0,0 +1,250 @@
+
+        ProjectNode newZ = project(rewrittenW);
+        ProjectNode newY = project(newZ);
+
+        memo.replace(yGroup, newY, "rule");
+
+        assertEquals(memo.getGroupCount(), 4);
+
+        assertMatchesStructure(
+                memo.extract(),
+                project(x.getId(),
+                        project(newY.getId(),
+                                project(newZ.getId(),
+                                        values(w.getId())))));
+    }
+
+    /*
+      From: X -> Y -> Z
+      To:   X -> Z
+     */
+    @Test
+    public void testRemoveNode()
+            throws Exception
+    {
+        PlanNode z = values();
+        PlanNode y = filter(z);
+        PlanNode x = project(y);
+
+        Memo memo = new Memo(idAllocator, x);
+
+        assertEquals(memo.getGroupCount(), 3);
+
+        int yGroup = getChildGroup(memo, memo.getRootGroup());
+        memo.replace(yGroup, memo.getNode(yGroup).getSources().get(0), "rule");
+
+        assertEquals(memo.getGroupCount(), 2);
+
+        assertMatchesStructure(
+                memo.extract(),
+                project(x.getId(),
+                        values(z.getId())));
+    }
+
+    /*
+       From: X -> Z
+       To:   X -> Y -> Z
+     */
+    @Test
+    public void testInsertNode()
+            throws Exception
+    {
+        PlanNode z = values();
+        PlanNode x = project(z);
+
+        Memo memo = new Memo(idAllocator, x);
+
+        assertEquals(memo.getGroupCount(), 2);
+
+        int zGroup = getChildGroup(memo, memo.getRootGroup());
+        PlanNode y = filter(memo.getNode(zGroup));
+        memo.replace(zGroup, y, "rule");
+
+        assertEquals(memo.getGroupCount(), 3);
+
+        assertMatchesStructure(
+                memo.extract(),
+                project(x.getId(),
+                        filter(y.getId(),
+                                values(z.getId()))));
+    }
+
+    /*
+      From: X -> Y -> Z
+      To:   X --> Y1' --> Z
+              \-> Y2' -/
+     */
+    @Test
+    public void testMultipleReferences()
+            throws Exception
+    {
+        PlanNode z = values();
+        PlanNode y = filter(z);
+        PlanNode x = project(y);
+
+        Memo memo = new Memo(idAllocator, x);
+        assertEquals(memo.getGroupCount(), 3);
+
+        int yGroup = getChildGroup(memo, memo.getRootGroup());
+
+        PlanNode rewrittenZ = memo.getNode(yGroup).getSources().get(0);
+        PlanNode y1 = filter(rewrittenZ);
+        PlanNode y2 = project(rewrittenZ);
+
+        PlanNode newX = join(y1, y2);
+        memo.replace(memo.getRootGroup(), newX, "rule");
+        assertEquals(memo.getGroupCount(), 4);
+
+        assertMatchesStructure(
+                memo.extract(),
+                join(newX.getId(),
+                        filter(y1.getId(), values(z.getId())),
+                        project(y2.getId(), values(z.getId()))));
+    }
+
+    private static void assertMatchesStructure(PlanNode actual, PlanNode expected)
```

Comment:
I think it would be better to use `PlanMatchPattern` here. I don't think it is a good idea to create two frameworks for the same.